scipy-stubs/signal/windows/_windows.pyi

```diff
@@ -29,36 +35,179 @@ __all__ = [
-def triang(M: Untyped, sym: bool = True) -> Untyped: ...
-def parzen(M: Untyped, sym: bool = True) -> Untyped: ...
-def bohman(M: Untyped, sym: bool = True) -> Untyped: ...
-def blackman(M: Untyped, sym: bool = True) -> Untyped: ...
-def nuttall(M: Untyped, sym: bool = True) -> Untyped: ...
-def blackmanharris(M: Untyped, sym: bool = True) -> Untyped: ...
-def flattop(M: Untyped, sym: bool = True) -> Untyped: ...
-def bartlett(M: Untyped, sym: bool = True) -> Untyped: ...
-def hann(M: Untyped, sym: bool = True) -> Untyped: ...
-def tukey(M: Untyped, alpha: float = 0.5, sym: bool = True) -> Untyped: ...
-def barthann(M: Untyped, sym: bool = True) -> Untyped: ...
-def general_hamming(M: Untyped, alpha: Untyped, sym: bool = True) -> Untyped: ...
-def hamming(M: Untyped, sym: bool = True) -> Untyped: ...
-def kaiser(M: Untyped, beta: Untyped, sym: bool = True) -> Untyped: ...
-def kaiser_bessel_derived(M: Untyped, beta: Untyped, *, sym: bool = True) -> Untyped: ...
-def gaussian(M: Untyped, std: Untyped, sym: bool = True) -> Untyped: ...
-def general_gaussian(M: Untyped, p: Untyped, sig: Untyped, sym: bool = True) -> Untyped: ...
-def chebwin(M: Untyped, at: Untyped, sym: bool = True) -> Untyped: ...
-def cosine(M: Untyped, sym: bool = True) -> Untyped: ...
-def exponential(M: Untyped, center: Untyped | None = None, tau: float = 1.0, sym: bool = True) -> Untyped: ...
-def taylor(M: Untyped, nbar: int = 4, sll: int = 30, norm: bool = True, sym: bool = True) -> Untyped: ...
+_Array_f8_1d: TypeAlias = np.ndarray[tuple[int], np.dtype[np.float64]]
+_Array_f8_2d: TypeAlias = np.ndarray[tuple[int, int], np.dtype[np.float64]]
+
+_Norm: TypeAlias = Literal[2, "approximate", "subsample"]
+_WindowLength: TypeAlias = int | np.int16 | np.int32 | np.int64
+_Window: TypeAlias = Literal[
+    "barthann",
+    "brthan",
+    "bth",
+    "bartlett",
+    "bart",
+    "brt",
+    "blackman",
+    "black",
+    "blk",
+    "blackmanharris",
+    "blackharr",
+    "bkh",
+    "bohman",
+    "bman",
+    "bmn",
+    "boxcar",
+    "box",
+    "ones",
+    "rect",
+    "rectangular",
+    "cosine",
+    "halfcosine",
+    "exponential",
+    "poisson",
+    "flattop",
+    "flat",
+    "flt",
+    "hamming",
+    "hamm",
+    "ham",
+    "hann",
+    "han",
+    "lanczos",
+    "sinc",
+    "nuttall",
+    "nutl",
+    "nut",
+    "parzen",
+    "parz",
+    "par",
+    "taylor",
+    "taylorwin",
+    "triangle",
+    "triang",
+    "tri",
+    "tukey",
+    "tuk",
+]
+_WindowNeedsParams: TypeAlias = Literal[
+    "chebwin",
+    "cheb",
+    "dpss",
+    "gaussian",
+    "gauss",
+    "gss",
+    "general cosine",
+    "general_cosine",
+    "general gaussian",
+    "general_gaussian",
+    "general gauss",
+    "general_gauss",
+    "ggs",
+    "general hamming",
+    "general_hamming",
+    "kaiser",
+    "ksr",
+    "kaiser bessel derived",
+    "kbd",
+]
+
+def general_cosine(M: _WindowLength, a: Sequence[AnyReal], sym: op.CanBool = True) -> _Array_f8_1d: ...
```

Comment:
The docs say that `a` should accept array-likes here, but `numpy.ndarray` isn't assignable to `Sequence`. So you could use `numpy._typing._ArrayLikeFloat_co` for this; `Sequence[AnyReal]` and `numpy.ndarray` are both assignable to it.

Alternatively, if you want to enforce `a` to be 1-d, you could keep the `Sequence[AnyReal]`, and union it with `numpy._typing._ArrayLike[np.floating[Any] | np.integer[Any] | np.bool_]`. The latter can be any shape, but at least the `Sequence` part will is enforced as 1d.

Reply:
The code I believe conflicts with the docs here because the body is
```python
def general_cosine(M, a, sym=True):
    if _len_guards(M):
        return np.ones(M)
    M, needs_trunc = _extend(M, sym)

    fac = np.linspace(-np.pi, np.pi, M)
    w = np.zeros(M)
    for k in range(len(a)):
        w += a[k] * np.cos(k * fac)

    return _truncate(w, needs_trunc)
```
so I don't think `ArrayLike` works here because no where in the code does it convert the array like into an array. 

I defaulted to `Sequence` because you have to be able to:
- call `len`
- index the array and get some scalar out

The incompatibility with `np.ndarray` was something I forgot about though. This would crash if given a numpy scalar even though that is an array like we would need something like `Sequence[AnyReal]` unioned with the allowed ndarray types.

Reply:
Is there another type alias that is basically a protocol which requires a `__len__` and `__getitem__` implementation? Feel like that is moreso the correct type, although maybe not the intention due to the docs (even if it conflicts with the function implementation).

Reply:
Ah good catch! Then maybe `Sequence[AnyReal] | npt.NDArray[floating[Any]]` 🤔 ?

Reply:
Because the annotations of `numpy.ndarray.__getitem__` currently aren't very good, so I don't think it will work if you'd only use e.g. `optype.CanGetitem` here.

Reply:
I added 1d ndarray types for all the numpy scalar types allowed by `AnyReal` so this should be fixed now.